pkg/discovery/dns/resolver.go

```diff
@@ -71,7 +78,13 @@ func (s *dnsSD) Resolve(ctx context.Context, name string, qtype QType) ([]string
 		}
 		ips, err := s.resolver.LookupIPAddr(ctx, host)
 		if err != nil {
-			return nil, errors.Wrapf(err, "lookup IP addresses %q", host)
+			// https://github.com/thanos-io/thanos/issues/3186
+			// Default DNS resolver can make thanos components crash if DNS resolutions results in EAI_NONAME.
```

Comment:
This comment does not make sense, to me. Why telling a story in some deep Thanos package what happened? There might be more consumers of this than just  Thanos (e.g Cortex).

Let's focus on what user of this package can expect from the piece. And let's comment only if something is surprising. (:

Reply:
I would only say:

```
We exclude error from std Golang resolver for the case of the domain (e.g `NXDOMAIN`) not being found by DNS server. Since `miekg` does not consider this as an error,  when the host cannot be found, empty slice will be returned.
```

WDYT?